remotefs/downloaddirectory.go

```diff
@@ -0,0 +1,49 @@
+package remotefs
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+// DownloadDirectory downloads all files and directories recursively from the remote system to local directory.
+func DownloadDirectory(fs FS, src, dst string) error {
+	// Ensure the destination directory is created even if empty
+	srcInfo, err := fs.Stat(src)
+	if err != nil {
+		return fmt.Errorf("stat dir for download: %w", err)
+	}
+	mode := srcInfo.Mode() & os.ModePerm
+	if err := os.MkdirAll(dst, mode); err != nil {
+		return fmt.Errorf("create empty dir for download: %w", err)
+	}
+
+	entries, err := fs.ReadDir(src)
```

Comment:
Did you try with `io/fs.WalkDir`? I think it could simplify this a bit. Something like:

```go
func DownloadDirectory(fsys fs.FS, src, dst string) error {
	return fs.WalkDir(fsys, src, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		
		relPath, err := filepath.Rel(src, path)
		if err != nil {
			return fmt.Errorf("calculate relative path: %w", err)
		}
		targetPath := filepath.Join(dst, relPath)

		if d.IsDir() {
			return os.MkdirAll(targetPath, os.ModePerm)
		}

		fileInfo, err := d.Info()
		if err != nil {
			return fmt.Errorf("get file info: %w", err)
		}

		return Download(fsys, path, filepath.Join(targetPath, fileInfo.Name()))
	})
}
```

WalkDir does not dive into symlinks, but that may be a good thing.

Reply:
I haven't, but that's a good idea, I have rewrote it like that. had to wrap errors to satisfy the linter. 
IMO we don't mind if it doesn't process symlinks(it could be quite weird and unexpected if it did)